src/Aspire.Hosting/Pipelines/DistributedApplicationPipeline.cs

```diff
@@ -154,7 +164,67 @@ private static (List<PipelineStep> StepsToExecute, Dictionary<string, PipelineSt
         }
 
         var stepsToExecute = ComputeTransitiveDependencies(targetStep, allStepsByName);
-        stepsToExecute.Add(targetStep);
+
+        // Also include all steps that are transitively required by the target step
+        var requiredBySteps = ComputeTransitiveRequiredBy(targetStep, allStepsByName);
+        foreach (var step in requiredBySteps)
+        {
+            if (!stepsToExecute.Contains(step))
+            {
+                stepsToExecute.Add(step);
+            }
+
+            // For each step that's required by the target step, also include its dependencies
+            var stepDependencies = ComputeTransitiveDependencies(step, allStepsByName);
+            foreach (var dependency in stepDependencies)
+            {
+                if (!stepsToExecute.Contains(dependency))
+                {
+                    stepsToExecute.Add(dependency);
+                }
+            }
+        }
+
+        // Add the target step if not already present
+        if (!stepsToExecute.Contains(targetStep))
+        {
+            stepsToExecute.Add(targetStep);
+        }
+
+        // Additional pass: Include steps required by ANY executing step
```

Comment:
This feels broken to me. It should only need 1 pass.

Reply:
Yeah, this is machinations of late-night Safia but I think this will look cleaner once we do the follow-up PR and add root nodes for "build" and "deploy compute" to the model.